guikit/threads.py

```diff
@@ -196,17 +215,43 @@ def abort_thread(self, ident: int) -> None:
         Raises:
             KeyError: If the thread identifier is not in the ThreadPool
         """
-        try:
-            self._workers[ident].abort = True
-        except KeyError as key_err:
-            key_err_ = KeyError(f"Thread with index: {ident} is not in the ThreadPool.")
-            logger.exception(key_err_)
-            raise key_err_ from key_err
+        self.get_thread(ident).abort = True
+
+    def join_thread(self, ident: int) -> None:
+        """Join the thread specified by `ident`.
+
+        Args:
+            ident: Thread identifier
+
+        Raises:
+            KeyError: If the thread identifier is not in the ThreadPool
+        """
+        self.get_thread(ident).join()
 
     def post_event(self, event: ThreadResult):
         """Adds an event to the event loop of the main thread."""
         wx.PostEvent(self._window, event)
 
+    def stop_threads(self):
+        """Stop all worker threads and wait for them to finish.
+
+        Todo:
+            Leave daemon threads running?
+            Race condition if a thread is created while this function is running
+        """
+        # Copy the list of worker threads in case the running threads try to
+        # modify self._workers
+        with self._workers_lock:
+            workers = self._workers.values()
+
+        # Tell workers that we want to abort
+        for worker in workers:
+            worker.abort = True
+
+        # Wait for all workers to finish
+        for worker in workers:
+            worker.join()
```

Comment:
I've been wondering if this is a sensible thing to do:
```suggestion
            if not worker.daemon:
                worker.join()
```

That way, daemon threads will still terminate when the main window is closed, but we don't wait to close the window for e.g. iPANACEA's data acquisition thread to finish. I think this is what we want: for a GUI app it makes sense to end all the background threads when the window disappears.

What do you think?

Reply:
It makes sense but, to be honest, it depends on what the daemon thread is doing. If it requires the main window for some reason - eg. it detects if there are updates for the software and shows a warning in the bottom left corner of the main window - then, things will fail if they are left to their own devices after the window is closed. I think it is cleaner for an app to properly deal with its threads, even if it is perfectly safe to leave them on their own.

Reply:
Okey doke. You're right that leaving it as is will be less bug prone!